core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+package textile
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/FleekHQ/space-daemon/log"
+
+	"github.com/textileio/go-threads/core/thread"
+)
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) BackupFileWithReader(ctx context.Context, bucket Bucket, path string, reader io.Reader) (err error) {
+
+	if _, _, err = tc.UploadFileToHub(ctx, bucket, path, reader); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if _, err = tc.setMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error setting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) backupFile(ctx context.Context, bucket Bucket, path string) error {
+
+	errc := make(chan error, 1)
+	pipeReader, pipeWriter := io.Pipe()
+
+	// go routine for piping
+	go func() {
+		defer close(errc)
+		defer pipeWriter.Close()
+
+		err := bucket.GetFile(ctx, path, pipeWriter)
+		if err != nil {
+			errc <- err
+			log.Error(fmt.Sprintf("error getting file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+			return
+		}
+	}()
+	if err := <-errc; err != nil {
+		return err
+	}
+
+	if err := tc.BackupFileWithReader(ctx, bucket, path, pipeReader); err != nil {
+		log.Error(fmt.Sprintf("error backing up file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup all files in a bucket
+func (tc *textileClient) backupBucketFiles(ctx context.Context, bucket Bucket, path string) (int, error) {
+	var wg sync.WaitGroup
+	var count int
+
+	// XXX: we ignore errc (no atomicity at all) for now but we should return
+	// XXX: the errors, perhaps in a separate async call
+	errc := make(chan error)
+
+	dir, err := bucket.ListDirectory(ctx, path)
+	if err != nil {
+		return 0, err
+	}
+
+	wg.Add(len(dir.Item.Items))
+
+	for _, item := range dir.Item.Items {
+		if item.IsDir {
+			p := strings.Join([]string{path, item.Name}, "/")
+
+			n, err := tc.backupBucketFiles(ctx, bucket, p)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+			continue
+		}
+
+		// parallelize the backups
+		go func(path string) {
+			defer wg.Done()
+
+			if err = tc.backupFile(ctx, bucket, path); err != nil {
+				errc <- err
```

Comment:
What happens if there's no one listening at `errc`? I've seen Go get stuck in those cases, that's why I've had to do stuff like 
```go
select {
  case errc <- err:
    log.Debug("Msg sent")
  default:
    // Do nothing
}
```

Maybe since this is within a go subroutine it doesn't matter but I'd make sure it doesn't hang when for example `Ctrl+C`ing the daemon.

Reply:
I don't think select is necessary in this case @dmerrill6, since in this case we are sending `err` through the `errc` and not waiting for a value from `errc`.Also, it wouldn't block because of the same reason.

Reply:
@perfectmak 
@dmerrill6 

It is a buffered channel (`make(chan, 1)`), so it blocks only when the buffer is full (ie: in this case if `err` is sent but not received.)

Reply:
@dmerrill6 @perfectmak 

I've added the `select` to avoid any confusion in https://github.com/FleekHQ/space-daemon/pull/191/files/593541272f2e1edb98679e8020351e5a6ecb5232#diff-b580f3c0539be5a72155b66b76313a90R103 (Please ignore my previous comment about the buffered channel.)